CIP-????/README.md

```diff
@@ -0,0 +1,334 @@
+
+## Specification
+
+### Validation zones
+
+A validation zone is a container for transactions. 
+The main property of a validation zone is that the node may need to look at all the transactions in the zone in order to decide if any of them are valid. 
+Within validation zones we can therefore allow features that require “earlier” transactions to depend on “later” transactions for their validity.
+
+At the ledger level, a validation zone:
+
+- Is a new top-level element (an alternative to a transaction)
+- Contains a sequence of transactions (not zones, there is no point nesting them)
+- Carries no further data beyond this
+- Triggers a special mode of validation for the transactions within (see “validating a zone” below)
+
+### Unresolved inputs and resolving outputs
+
+We add a new kind of transaction input, an “unresolved input” (UTXI). 
+An unresolved input differs from a normal transaction input:
+
+- It does not specify a transaction output that it spends
+- It does specify the value that the input provides
+- It has a credential and an optional datum
+
+Unresolved inputs must be addressable. 
+Current transaction inputs are not addressable by the combination of transaction ID and an index, since they are kept in a set. 
+We therefore need to implement https://github.com/cardano-foundation/CIPs/pull/758, or to keep the unresolved inputs in a separate list.
+
+We add a new kind of transaction output, a “resolving output”. 
+A resolving output differs from a normal transaction output:
+- It specifies that it “resolves” a particular unresolved transaction input, by specifying the transaction ID and index of the input
+- It does not specify the value that it provides
+- It does not have a credential
+
+Unresolved inputs and resolving outputs are illegal outside a validation zone. 
+They are included in the script context accurately, with one subtlety: in the same way that normal transaction inputs are decorated with the value that they provide, resolving transaction outputs are similarly decorated with the value that they provide.
+
+### Required transactions
+
+We add a new field to the transaction body, `requiredTxs`, which is a list of transaction IDs.
+
+The `requiredTxs` field is special in that it is _not_ included when hashing the transaction body to create the transaction ID. 
+It is however signed when the transaction body is signed.
+
+The required transactions field is illegal outside a validation zone. 
+It is included in the script context.
+
+### Validating a zone
+
+Validating a zone proceeds as follows:
+
+#### 1: Check all zone-level properties
+The following zone-level properties are checked:
+
+1. (UTXIs are resolved) The unresolved inputs and resolving outputs in the zone match up exactly.
+2. (Linearizability) The transactions are linearizable, that is: there is a reordering of the transactions in the zone such that every resolving output comes before every unresolved input which it resolves.
```

Comment:
they do not actually have to be ordered, right? It should just be possible?

Reply:
What's the benefit here? If you just require the zone to be built in the linearized order, the node doesn't have to do the special check, and offchain code doesn't have to do a search for that order.

Reply:
> If you just require the zone to be built in the linearized order, the node doesn't have to do the special check

Well, it will still have to check that it _is_ in order as it goes, but indeed I think this is a good idea! The whole point of the "zone" as an explicit construct is to tell the node more explicitly what is happening so it has to do less work. Also pre-computing the order is another good move in that direction.

(I think this is a remnant of when we were thinking of people actually submitting these transactions to the node out-of-order, with the node reassembling them into the zone. But we're not going that way at the moment.)

> and offchain code doesn't have to do a search for that order

Why would it need to do that? 

Also, requiring the zone to be submitted in order will require the off-chain _submitting_ code to find an order. But that's not too bad.

Reply:
The offchain submitting code finds the order by construction, sure; but when the downstream indexer then processed the validation zone, it'll loop through the transactions in order to update its state / materialized view. If the zone was out of order, it'd have to do extra processing (at the very least multiple passes to build up a list of satisfying outputs and requested inputs) before actually processing each tx.

If it's already sorted such that it can process the state changes in order, and it can rely on that, then the code becomes a lot simpler for the N downstream consumers, at the expense of more complexity for the one upstream producer. Especially when the upstream producer likely already has to do that work to know the TX is valid before submitting anyway.

(Ultimately would also be solved by having the mini protocols emit in the linearized order, or having Ogmios etc. do that transformation, etc)

Reply:
because transactions have only a partial order, i think the ordering and checking algorithm will both be O(same number of comparisons). However, there is the usual constraint on the order : a **normal** input has to come after the **normal** output it is spending. Same goes for the **weird** ones, but backwards. The linearizability check still has to do a lot of comparisons to make sure that any two transactions dont have **both** a weird **and** a normal dependency.

Reply:
> The linearizability check still has to do a lot of comparisons (O(n^2) ?) to make sure that any two transactions dont have both a weird and a normal dependency.

This is just cycle detection so it should be O(V+E) at worst.

Reply:
Ah, I realised a complexity. The "linearized" order is not actually the order that makes checking the most straightforward. The reason is that value is specified on the _UTXI_. So when processing a transaction with a resolving output, you really want to have seen the corresponding UTXI already otherwise you don't know if it balances.

So the order for easiest transaction-by-transaction checking is one that satisfies the reverse condition: every UTXI comes before the output that resolves it.

Putting it another way, the _information_ flows in the opposite direction to the _value_.

I think this suggests that the zone should be ordered so that UTXIs come before their resolving outputs, and we keep the separate linearizability check. Alternatively, we could generate some extra state in the initial checking phase so we know about all the UTXIs up-front?